tensorflow_model_optimization/g3doc/guide/clustering/clustering_with_keras.ipynb

```diff
@@ -0,0 +1,1164 @@
+    "\n",
+    "#### Why is this useful?\n",
+    "\n",
+    "Clustering ensures that only a specified number of unique values (cluster centroids) are used in a weight tensor. This allows for certain types of hardware to benefit from advanced weight compression techniques. \n",
+    "\n",
+    "#### How does it work?\n",
+    "\n",
+    "The clustering API is designed to fine tune the weights after certain training accuracy is achieved. It is implemented using a lookup table to hold the cluster centroids during model fine tuning. The weight array is populated with 'gather' operation so that during back propagation the gradients can be calculated in a normal way. The lookup table is then adjusted using the cumulative gradient values for the weights that correspond to the same centroid. For more details on the usage of the API, please refer to the [GitHub page]().\n",
+    "\n",
+    "In this tutorial, we will walk you through an end-to-end example of using the clustering API on a simple sequential model trained on the MNIST dataset. By using a generic file compression algorithm (e.g. zip) the Keras model will be reduced in size, and the size reduction persists when converted to a TensorFlow Lite format.\n",
+    "\n",
+    "Similar to the [pruning API](https://www.tensorflow.org/model_optimization/guide/pruning/pruning_with_keras), the technique and API are not TensorFlow Lite specific. We just show its application on the TensorFlow Lite backend, as it covers size-sensitive use-cases.\n",
+    "\n",
+    "In the tutorial we will:\n",
+    "\n",
+    "1.   Train a simple model with the MNIST dataset with Keras from scratch.\n",
+    "2.   Train a clustered model with the MNIST dataset with the clustering API.\n",
+    "3.   Compare the size of the clustered model and the non-clustered one after compression.\n",
+    "4.   Convert the clustered model to TensorFlow Lite format and verify the accuracy.\n",
+    "5.   Show how the clustered model works with other optimization techniques, for instance, post-training quantization.\n",
+    "\n",
+    "\n",
+    "## Setup\n",
+    "\n",
+    "To use the clustering API, install the `tensorflow-model-optimization` package. See the [TensorFlow model optimization repo](https://github.com/tensorflow/model-optimization) for compatible API versions.\n",
+    "\n",
+    "Since you will train a few models in the tutorial, install the `tensorflow-gpu` package to speed things up. Enable the GPU with: *Runtime > Change runtime type > Hardware accelator* and make sure GPU is selected."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "! pip uninstall -y tensorflow\n",
```

Comment:
Should we include all the "pip install" set up at the beginning of the file? Will it disrupt the user's local environment? or are we assuming everyone who uses this jupyter notebook will run it in colab?

Reply:
Resolving since discussed previously. The assumption is that people have followed the installation instructions -> they will use virtualenv or colab, though we could have a comment with a pointer to them